### PR TITLE
first attempt at graphics layering

### PR DESCRIPTION
**Pull request by ExtraGravity**
This change is [Reviewable](https://reviewable.io/reviews/enochtsang/terminalfighter/72)

### CODE REVIEW

File: src/graphics_handler.cpp [hunk 1]
```diff
@@ -1,22 +1,48 @@
...
 }
 
-void GraphicsHandler::draw(SDL_Texture *texture, int x_pos, int y_pos) {
+void GraphicsHandler::draw(SDL_Texture *texture, int x_pos, int y_pos, GraphicPriority priority) {
     SDL_Rect texture_rect;
     SDL_QueryTexture(texture, NULL, NULL, &texture_rect.w, &texture_rect.h);
     texture_rect.x = x_pos;
     texture_rect.y = y_pos;
-    SDL_RenderCopy(renderer_, texture, NULL, &texture_rect);
+    this->draw(texture, texture_rect, priority);
```
Comment (bseto) [inline src/graphics_handler.cpp:35]: Pretty good! I like the use of `this` here.

File: src/draw_request.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include <memory>
```
Comment (bseto) [inline src/draw_request.h]: Where is this used?

File: src/missile.cpp [hunk 3]
```diff
@@ -18,7 +18,7 @@ void Missile::draw(GraphicsHandler *graphics) {
         printf("Missile graphics were null! Setting missile graphic...\n");
         set_texture(graphics->load_image("assets/projectiles/missile.png"));
     }
-    graphics->draw(missile_texture_, (int)x_pos(), (int)y_pos());
+    graphics->draw(missile_texture_, (int)x_pos(), (int)y_pos(), GraphicPriority::MIDDLE);
```
Comment (bseto) [inline src/missile.cpp:21]: I like the `GraphicPriority` here. But just as a note: You wanted to know what the next `TODO` should be. Maybe we should work on doing that global config file soon? 
That way we can have the `GraphicPriority::MIDDLE` be something like `GlobalConfig.get(PriorityMissile)` or something along those lines. 

That way, we'll be able to have a config file that we can edit to change all of their layers.

**Review by bseto — COMMENTED** [review]
Pretty good so far from my first read. I will go into the details tomorrow morning!

File: src/draw_request.h (hunk 2 above)
Comment (ExtraGravity) [reply src/draw_request.h]: I was trying to use smart pointers before, but I realize now it's not necessary. I'll delete this line.

File: src/missile.cpp (hunk 3 above)
Comment (ExtraGravity) [reply src/missile.cpp:21]: I was thinking, that maybe we should push for something playable first. Then do a big cleanup like adding using a config file, but having something playable first might help motivate us better.

Comment (bseto) [reply src/missile.cpp:21]: ok, that sounds good

File: src/draw_request.h [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    
+    GraphicPriority z_level() const {
+        return z_level_;
+    }
+
+
+private:
+    SDL_Texture* texture_;
+    SDL_Rect texture_rect_;
+    GraphicPriority z_level_;
```
Comment (bseto) [inline src/draw_request.h]: Is there a point to store the z_level_ inside GraphicPriority?

The `void GraphicsHandler::draw(SDL_Texture *texture, int x_pos, int y_pos, GraphicPriority priority)` function has `draw_queue_.at(priority).push_back(DrawRequest(texture, texture_rect, priority));`

which stores the DrawRequest in a vector according to it's priority already right?

I'm mostly wondering if you think we will ever have to query for the z_level_ in the future.

File: src/graphics_handler.cpp [hunk 5]
```diff
@@ -1,22 +1,48 @@
...
 }
 
 void GraphicsHandler::update_screen() {
+    for(auto priority: DRAW_ORDER) {
+        for(auto const& draw_request: draw_queue_[priority]) {
+            SDL_RenderCopy(renderer_, draw_request.texture(), NULL, &draw_request.texture_rect());
+        }
+        draw_queue_.at(priority) = std::vector<DrawRequest>();
```
Comment (bseto) [inline src/graphics_handler.cpp:43]: Very good! I was looking around to make sure you were clearing the old stuff to avoid memory leaks.

**Comment by bseto** [comment]
I think everything looks good. You just have to take out the `include <memory>`

File: src/draw_request.h (hunk 2 above)
Comment (bseto) [reply src/draw_request.h]: Completed.

File: src/draw_request.h (hunk 4 above)
Comment (bseto) [reply src/draw_request.h]: Completed.

**Comment by bseto** [comment]
Okay. Everything is done. All you gotta do is fix conflicts. 
- [x] Remove `include <memory>` header
- [x] Remove/explain `GraphicPriority z_level_;`
- [ ] Fix conflict

EDIT: Checklists like the above aren't bad on the meantime for tracking what's done haha...
reference: https://github.com/blog/1375-task-lists-in-gfm-issues-pulls-comments

**Pull request merged**